zcash_client_backend/src/data_api/testing/pool.rs

```diff
@@ -238,3 +256,2321 @@ pub fn send_single_step_proposed_transfer<T: ShieldedPoolTester>(
+        .usk()
+        .transparent()
+        .derive_secret_key(Scope::External.into(), default_index)
+        .unwrap();
+    let outpoint = OutPoint::fake();
+    let txout = TxOut {
+        script_pubkey: default_addr.script(),
+        value,
+    };
+    // Add the fake input to our UTXO set so that we can ensure we recognize the outpoint.
+    st.wallet_mut()
+        .put_received_transparent_utxo(
+            &WalletTransparentOutput::from_parts(outpoint.clone(), txout.clone(), None).unwrap(),
+        )
+        .unwrap();
+
+    assert_matches!(builder.add_transparent_input(sk, outpoint, txout), Ok(_));
+    let test_prover = LocalTxProver::bundled();
+    let build_result = builder
+        .build(
+            OsRng,
+            &test_prover,
+            &test_prover,
+            &zip317::FeeRule::standard(),
+        )
+        .unwrap();
+    let txid = build_result.transaction().txid();
+    st.wallet_mut()
+        .store_decrypted_tx(DecryptedTransaction::<DSF::AccountId>::new(
+            None,
+            build_result.transaction(),
+            vec![],
+            #[cfg(feature = "orchard")]
+            vec![],
+        ))
+        .unwrap();
+
+    // Verify that storing the fully transparent transaction causes a transaction
+    // status request to be generated.
+    let tx_data_requests = st.wallet().transaction_data_requests().unwrap();
+    assert!(tx_data_requests.contains(&TransactionDataRequest::GetStatus(txid)));
+
+    // We call get_wallet_transparent_output with `allow_unspendable = true` to verify
+    // storage because the decrypted transaction has not yet been mined.
+    let utxo = st
+        .wallet()
+        .get_all_unspent_transparent_output(&OutPoint::new(txid.into(), 0), true)
+        .unwrap();
+    assert_matches!(utxo, Some(v) if v.value() == utxo_value);
+
+    // That should have advanced the start of the gap to index 11.
+    let new_known_addrs = st
+        .wallet()
+        .get_known_ephemeral_addresses(account_id, None)
+        .unwrap();
+    assert_eq!(new_known_addrs.len(), (GAP_LIMIT as usize) + 11);
+    assert!(new_known_addrs.starts_with(&known_addrs));
+
+    let reservation_should_succeed = |st: &mut TestState<_, DSF::DataStore, _>, n| {
+        let reserved = st
+            .wallet_mut()
+            .reserve_next_n_ephemeral_addresses(account_id, n)
+            .unwrap();
+        assert_eq!(reserved.len(), n);
+        reserved
+    };
+    let reservation_should_fail =
+        |st: &mut TestState<_, DSF::DataStore, _>, n, _expected_bad_index| {
+            assert_matches!(
+                st.wallet_mut()
+                    .reserve_next_n_ephemeral_addresses(account_id, n),
+                Err(..)
+            );
+            // previously it was hard-coded to a SqlClientError which could check the index
+            // not sure how best to handle this in a generic way
+            // Err(SqliteClientError::ReachedGapLimit(acct, bad_index))
```

Comment:
Any thoughts on how to handle checking for specific errors like this in the back-end agnostic case?